File: app/models/request.rb
class Request < ApplicationRecord
  validates_presence_of   [:lat, :lng, :area, :phone_number, :material]
  belongs_to        :material
  belongs_to        :assigned_partner, class_name: 'Partner', foreign_key: :assigned_id, optional: true

  enum status: %i(created assigned)

  def api_response
    response = {
      id: self.id,
      material: {
        id: self.material.id,
        name: self.material.name
      },
      area: self.area,
      status: self.status,
    }

    response[:assigned_partner] = {
      id: self.assigned_partner.id,
      name: self.assigned_partner.name
    } unless self.assigned_partner.blank?

    response
  end
end

File: app/javascript/packs/app/models/request.ts
import { Material } from "./material";
import { Partner } from "./partner";

export enum RequestStatus {
  Created = 0,
  Assigned = 1
}

export interface Request {
  id: number
  material: Material
  area: number
  status: RequestStatus
  phone_number: string
  lat: number
  lng: number
  assigned_partner?: Partner
}